feat: add host availability strategy to allow hosts marked unavailable chances for retry (#530)
awslabs/aws-advanced-jdbc-wrapper: docs/using-the-jdbc-driver/HostAvailabilityStrategy.md

> # Host Availability Strategy

Reviewer: Can this page be added [here](https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/Documentation.md)?

Author: Fixed!

## What is Host Availability Strategy

> The AWS Advanced JDBC Driver keeps track of the availability of database hosts. A hosts availability status can be either `AVAILABLE` or `NOT_AVAILABLE`. This information is used by the driver when deciding which hosts to connect to.  

Reviewer: ```suggestion
The AWS JDBC Driver keeps track of the availability of database hosts. A host's availability status can be either `AVAILABLE` or `NOT_AVAILABLE`. This information is used by the driver when deciding which hosts to connect to.  
```
?

Author: Fixed!

> Host Availability Strategy helps the driver by allowing extra functionality to be added when a hosts availability is set or fetched. Different strategies can be swapped out for different behaviors, and are configurable as specified below.   

Reviewer: Extra functionality?

Author: Added some more details about what "extra functionality"

| Simple Host Availability Strategy                      | `null` or blank             | This strategy is a simple pass through strategy. It does nothing when availability is set, and returns the raw host availability when fetched.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 

> | Passive Exponential Backoff Host Availability Strategy | `passiveExponentialBackoff` | This strategy is passive by not actively health checking hosts. Instead it allows consumers of host availability check instead. <br/> If a host is marked `NOT_AVAILABLE`, there is a backoff time period where the hosts availability will return `NOT_AVAILABLE`. This backoff time period can be set using the `hostAvailabilityStrategyBaseBackoffTime` configuration parameter. After this period elapses, the hosts availability will return `AVAILABLE` so that consumers of this information may try the host out. <br/> Each subsequent time the host found to be down, the host is set to `NOT_AVAILABLE`, and another backoff period twice the duration of the last, will occur before the host availability starts to return `AVAILABLE`. If a host is found to be However, if the host is back up, consumers will set the host availability to `AVAILABLE` with no backoff period. The next backoff period that occurs will be reset to the base backoff time. | 

Reviewer 2: It's nt clear whether hostAvailabilityStrategyBaseBackoffTime in sec or millis.

# Host Availability Strategy

## What is Host Availability Strategy

> The AWS JDBC Driver keeps track of the availability of database hosts. A host's availability status can be either `AVAILABLE` or `NOT_AVAILABLE`. This information is used by the driver when deciding which hosts to connect to.  

Reviewer 3: I wonder if we should rename `NOT_AVAILABLE` to `UNAVAILABLE`

Author: I dont feel too strongly either way, but I think Im more used to seeing `UNAVAILABLE` being used.  
Sergiy mentioned `UNAVAILABLE` can look similar to `AVAILABLE`, so `NOT_AVAILABLE` is easier to read when quickly scanning the code. But  he asked to leave the decision in our hands. 
Do you feel particularly strongly either way?

Author: I just went ahead and renamed it. Done!

> Host Availability Strategies help the driver by adding extra functionality when a hosts availability is set or fetched. For example with the Exponential Backoff strategy, when a host's availability is set to `NOT_AVAILABLE` and after a backoff period, subsequent fetches of its availability will return `AVAILABLE` to allow the driver to retry the host. 

Reviewer: ```suggestion
When database hosts fail, they are marked as `NOT_AVAILABLE` and will not be checked again to see if they have recovered. Host Availability Strategies can help the driver keep host health statuses up to date by checking whether unavailable hosts have become available. For example, the Exponential Backoff strategy will let the driver retry hosts that have had their availability set to `NOT_AVAILABLE` by setting the host availability to `AVAILABLE` after a backoff period.
```
Maybe something like this?

Author: TY for the suggested changes. Done!

> Different strategies can be swapped out for different behaviors. The default Host Availability Strategy the driver uses can be user configured as specified below.   

Author: TY for the suggested changes. Done!

## Configuration Parameters
| Parameter                                 | Value     | Required | Description                                                                                                                            | Default Value |
|-------------------------------------------|-----------|----------|----------------------------------------------------------------------------------------------------------------------------------------|---------------|

> | `defaultHostAvailabilityStrategy`         | `String`  | No       | This is an override for the strategy that the driver will default to using. Possible values are listed below in the following section. | `null`        |

Reviewer 3: ```suggestion
| `defaultHostAvailabilityStrategy`         | `String`  | No       | This overrides the driver's default host availability strategy. Possible values are listed below in the following section. | `null`        |
```

Author: Done!

| `hostAvailabilityStrategyMaxRetries`      | `Integer` | No       | Maximum number of times a host availability strategy will retry a host that is not available.                                          | `5`           |

> | `hostAvailabilityStrategyBaseBackoffTime` | `Integer` | No       | For host availability strategies that employ a time-based backoff, this is the base time in seconds                                    | `30`          |

Reviewer 3: ```suggestion
| `hostAvailabilityStrategyBaseBackoffTime` | `Integer` | No       | The initial wait time in seconds. This parameter is only applicable for host availability strategies employing a time-based backoff.                                    | `30`          |
```

Author: Done!

> ## Host Availability Strategy Options

Reviewer: ```suggestion
## Host Availability Strategies
```

Author: Done!

|------------------------------------------------|----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | Simple Host Availability Strategy              | `null` or blank      | This strategy is a simple pass through strategy. It does nothing when availability is set, and returns the raw host availability when fetched.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | 

Reviewer 3: `null` or blank isn't very clear to me. What do you mean. If this is the default strategy I prefer we leave it out of the table and explicitly explain it below.

Author: Thats right. This is the default strategy. I'll be doing as you suggested.

Author: Done!

> | Simple Host Availability Strategy              | `null` or blank      | This strategy is a simple pass through strategy. It does nothing when availability is set, and returns the raw host availability when fetched.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | 

Reviewer: This is the first mention of raw host availability, can you elaborate on what that means?

Author: So raw host availability is just the host's availability without any extra logic or functionality. I will remove that wording and simply say that this strategy does not add any extra logic and functionality.

Author: Done!

> | Exponential Backoff Host Availability Strategy | `exponentialBackoff` | This strategy is passive by not actively health checking hosts. Instead it allows consumers of host availability check instead. <br/> If a host is marked `NOT_AVAILABLE`, there is a backoff time period where the hosts availability will return `NOT_AVAILABLE`. This backoff time period can be set using the `hostAvailabilityStrategyBaseBackoffTime` configuration parameter with the time units of seconds. After this period elapses, the hosts availability will return `AVAILABLE` so that consumers of this information may try the host out. <br/> Each subsequent time the host found to be down, the host is set to `NOT_AVAILABLE`, and another backoff period twice the duration of the last, will occur before the host availability starts to return `AVAILABLE`. If a host is found to be However, if the host is back up, consumers will set the host availability to `AVAILABLE` with no backoff period. The next backoff period that occurs will be reset to the base backoff time. | 

Reviewer: ```suggestion
| Exponential Backoff Host Availability Strategy | `exponentialBackoff` | This strategy does not actively health check hosts and instead allows host availability consumers to perform health checks. <br/><br/> If a host is marked `NOT_AVAILABLE`, there is a backoff time period where the host's availability will be `NOT_AVAILABLE`.  After the backoff period, the host's availability will be `AVAILABLE` so that host availability consumers can attempt to connect to the host. <br/><br/> Each subsequent time the host fails a health check, the host is set to `NOT_AVAILABLE` and another backoff period twice the duration of the last will occur before the host availability is set to `AVAILABLE` again. However, if the host passes the health check, consumers will set the host availability to `AVAILABLE` with no backoff period. The next backoff period that occurs will be reset to the base backoff time. <br/><br/> The backoff time period can be set in seconds using the `hostAvailabilityStrategyBaseBackoffTime` configuration parameter. | 
```
?

Author: Done!
